### ISSUE TITLE
Packages using `replace` are in need of moderating

### ISSUE THREAD

**Issue by peterjmit** [composer/packagist#362]
_TLDR;_ Requiring a package from a trusted source can result in receiving a package from an un-trusted source.

We already have #329 however this only addresses the issue from the point of a package maintainer/submitter, and not the end user

---

I had a situation this evening where by composer was pulling in an unexpected package:

My app had two dependencies relating to this issue (peterjmit/ecommerce-bundle is a private "VCS" package)

``` json
{
  "peterjmit/ecommerce-bundle": "~0.9",
  "stripe/stripe-php": "~1.8"
}
```

I was carrying out some upgrades, and switched stripe over to `"stripe/stripe-php": "~1.9"` however `peterjmit/ecommerce-bundle` had its own dependency which remained at `"stripe/stripe-php": "1.8.3"`

This mismatch in requirements resulted in composer downloading an unexpected package in its place

``` json
{
  "easybib/stripe-php": "dev-master"
}
```

I only discovered this because my test suite found that the class "Stripe" was not found.

Since this has happened I have been advised that this is due to the `replace` flag https://github.com/easybib/stripe-php/blob/master/composer.json#L35 for which there has now been a PR submitted to remove the `replace` entry. 

It is not impossible for a situation to arise (especially when taking into account #335) where malicious code could end up being incorporated into an application as a result of this behaviour. 

Therefore I think that packages containing `replace` either have to be heavily moderated, or perhaps better any repositories containing `replace` should require the original package owner to allow the flag.

Other options could include adding an option to composer to turn off the behaviour of `replace` however replace as described [in the composer docs](http://getcomposer.org/doc/04-schema.md#replace) is pretty useful. Or there could be a user prompt confirming the replacement.

As it stands I think that the fact that an end user can specify a package from a trusted entity (in this case Stripe) but receive a package from an unknown entity is dangerous. And while care & attention should be given to the requiring/updating of dependencies - under the current mechanism, this issue is easily missed.
